h2/src/main/org/h2/engine/Session.java

```diff
@@ -158,6 +160,7 @@
     private ArrayList<Value> temporaryLobs;
 
     private Transaction transaction;
+    private State state = State.INIT;
```

Comment:
what protects state? if nothing, does it not need to be volatile at least?

Reply:
That field is not alone, it's "protected" the same way as currentCommand or currentCommandStart. I mean, on mutation it's hopefully protected by a session lock (see discussion above), and on read Transaction.blockingTransaction is volatile, so the most important state "BLOCKED" will be determined correctly, but whether it's running or sleeping (or was slipping a nanosecond ago), I could care less, because volatile is not free.